### DISCUSSION TITLE
How to ensure no URL conflicts with collection routes

### DISCUSSION THREAD

**Discussion by mondomishi** [gatsbyjs/gatsby#32604] [Help]
# The Problem

- Suppose I've got a [collection route](https://www.gatsbyjs.com/docs/reference/routing/file-system-route-api#collection-routes) like `src/pages/products/{Product.name}.js`
- I want to make sure at build time that there are no two `Product`s with the same`.name` (because this would mean they would both get mapped to the same URL despite being different products).
- Is there any recommended way of doing this?

# What I've Tried

Specifically in my case, the url piece is coming from a yaml file in my repo. I've spent the last way-too-long trying to figure this out. Here's how far I got

- added an [`onPreExtractQueries`](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#onPreExtractQueries) to my `gatsby-node.js` which does something like this
    ```js
    // i chose this hook because it seems to run at all the important times that i would want to verify no url collisions
    exports.onPreExtractQueries = () => {
      const slugifiedProductNames = new Set();
      for (const file in myYamlFiles) {
        const contents = yaml.load(fs.readFileSync(file), 'utf8');
        const slugifiedProductName = slugify(contents.productName);
        if (slugifiedProductNames.has(urlPiece)) {
          throw new Error('oh no!');
        }
      }
    };
    ```
- the above is all well and good, but the problem is that the [slugify package is an ES module](https://github.com/sindresorhus/slugify/issues/10) (I am using this slugify package because [the gatsby docs](https://www.gatsbyjs.com/docs/reference/routing/file-system-route-api#routing-and-linking) tell me that is how gatsby produces the actual urls). When I try to import the `slugify` module in my `gatsby-node.js` I get this:
    ```
    Error in "<redacted_path_to_my_repo>/gatsby-node.js": Must use import to load ES Module: <redacted_path_to_my_repo>/node_modules/@sindresorhus/slugify/index.js
    ```
- i messed around with trying to get `esm` to work here, as mentioned in [this issue](https://github.com/gatsbyjs/gatsby/issues/7810#issuecomment-449741977) but ultimately never got anything to work
- I feel like the use case I am trying to solve for is not crazy. Am I missing some obvious + easier way to do this? I just want to make sure I don't unknowingly push colliding urls!

**Comment by herecydev** [comment]
I would definitely use your approach that you've gone with. What were the issues you had trying to use `esm`?

**Reply by mondomishi** [reply]
Howdy @herecydev, thanks for weighing in :)

When I follow the suggestion in [this comment](https://github.com/gatsbyjs/gatsby/issues/7810#issuecomment-449741977) I get the same error that I mentioned above:
```
Error in "<redacted_path_to_my_repo>/gatsby-node.js": Must use import to load ES Module: <redacted_path_to_my_repo>/node_modules/@sindresorhus/slugify/index.js
```
I also tried patching over just the single problematic require call
```
const yaml = require('js-yaml');
const fs = require('fs');
const path = require('path');
require = require('esm')(module);
const slugify = require('@sindresorhus/slugify');

// ...
```
(which again yields the same error)

Notably, the `slugify` package used by gatsby [is not transpiled](https://github.com/sindresorhus/slugify/issues/10), which I believe to be the cause of the error message in this situation

I'm curious, is my high level desire (ensuring no url conflicts) something that others are simply not testing against? In my mind this feels like something that would be a very common desire. So I'm confused that I am running into it and finding nothing.
